File: java/marker/src/main/java/com/github/wmacevoy/marker/Pencil.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.github.wmacevoy.marker;

/**
 *
 * @author wmacevoy
 */
public class Pencil implements WritingUtensil  {
    private Color color;
    
    public Pencil(Color color) {
        this.color = color;
    }

    public Color getColor() {
        return color;
    }

    public void setColor(Color color) {
        throw new UnsupportedOperationException();
    }
    
    public void draw() {
            System.out.println("sketching in " + getColor());
    }
    
    @Override
    public void write() throws BadMarkerState {
        draw();
    }
}

File: cpp/src/Pencil.cpp
#include "Pencil.h"

#include <iostream>

namespace marker {
    
        Pencil::Pencil(Color _color)
          : WritingUtensil(_color)
        {
            std::cout << "made " << getColor() << " pencil at " << (void*) (this) << std::endl;
        } 
        
        void Pencil::write() const {
            std::cout << "sketching in " << getColor() << std::endl;
        }

        std::shared_ptr<WritingUtensil> Pencil::clone() const {
            Pencil *copy = new Pencil(getColor());
            return std::shared_ptr<WritingUtensil>(copy);
        }

        Pencil::~Pencil() {
            std::cout << "destructing pencil@" 
                        << ((void*) this) << std::endl; 
        }
} // namespace marker